Send email to admin upon account request

johnsaigle opened the pull request:
## Brief summary of changes

When a user requests an account, an email will be sent to the study admin.

---

<img width="587" alt="Screen Shot 2020-02-18 at 16 50 38" src="https://user-images.githubusercontent.com/4022790/74781063-d0baba80-526e-11ea-9ce6-b51816616803.png">

---

#### Testing instructions (if applicable)

1. Follow the new steps added to the test plan in this PR.

#### Link(s) to related issue(s)

* Resolves #5943

driusan commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -223,6 +223,23 @@ class RequestAccount extends \NDB_Form
+            $config    = \NDB_Factory::singleton()->config();
+            $studyName = $config->getSetting('title');
+
+            \Email::send(
```
Before doing this you should ensure that the config setting is a valid email. I think it's also risky to do this by default, including sandboxes and places that haven't set up their email.. especially since that's not the purpose of the 'From' config setting and the default is 'no-reply@example.com'.

I think we should add a new config setting which defaults to empty, and only send an email if it's explicitly enabled.

johnsaigle replied:
Would this be two config settings then? e.g. `administratorEmail` and `emailAdminOnAccountRequest`?

driusan replied:
wouldn't something like `accountRequestContactEmail` be enough? If it's not set, there's no one to try emailing..

johnsaigle replied:
In #5965 I used the `From` setting in another context to test whether the server can send emails. Also in #6076 I did the same thing.

It's probably useful to have a separate email where an administrator can be reached if `From` is not suitable. I think 2 is better.

johnsaigle replied:
Blocked by #6101

driusan replied:
#6101 is merged

ridz1208 commented:
@johnsaigle I don't like the idea of having an email sent to one particular "admin" thats not how most LORIS instances work. User management is generally done per-sire or per-project.

Instead I would send a email to ALL users which have the site  AND the project the request was made to AND also have the user management permission in their permissions. On a properly managed LORIS project, that would probably only be the coordinator and their assistant maybe.

johnsaigle commented:
Thanks @ridz1208 that's a good idea. I'll add that.

ridz1208 commented:
@johnsaigle no longer blocked

johnsaigle commented:
@ridz1208 Just implemented it. Could you review?

ridz1208 commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -269,7 +269,67 @@ class RequestAccount extends \NDB_Form
 
+        // Verify reCAPTCHA
+        $err = [];
+        if (isset($values['g-recaptcha-response']) && isset($reCAPTCHAPrivate)) {
```
@johnsaigle why are the recptcha changes here ? bad rebase?

johnsaigle replied:
Yep. I'll fix it

ridz1208 commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -269,7 +269,67 @@ class RequestAccount extends \NDB_Form
+            $config    = \NDB_Factory::singleton()->config();
+            $studyName = $config->getSetting('title');
+
+            \Email::send(
```
I wont block this PR over this but I think we should use the notifier class in the future for this

driusan replied:
How much work is involved in changing it to use the Notifier class? That seems like a reasonable change..

ridz1208 replied:
not that much but it doesn't bring many benefits.. other I think than being able to subscribe or unsubscribe from email notifications

ridz1208 replied:
https://github.com/aces/Loris/wiki/Notification-system

johnsaigle replied:
That could be future work. You could make a ticket to refactor this to use the notification system

cmadjar replied:
@ridz1208 is there an issue already created for that? If not, would you mind creating it since you know what you are talking about and I don't? Thank you! :)

ridz1208 commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -269,7 +269,67 @@ class RequestAccount extends \NDB_Form
+        WHERE upr.CenterID=:center
+        AND uprojr.ProjectID=:project
+SQL;
+        $result = $DB->pselect(
+            $query,
+            ['center' => $site, 'project' => $project]
+        );
+        foreach ($result as $row) {
+            $hasPermission = \User::factory($row['userID'])
+                ->hasAnyPermission(
+                    ['user_accounts', 'user_accounts_multisite']
+                );
```
I find this is a weird combination of permission checking. you are doing SQL checks for site and project and User object checks for the module permissions. could you justify ? why not all SQL or all user class?

johnsaigle replied:
I would do all SQL but I'm not skilled enough with it to make a big query that does all of it. I'm also working everything out before my last day so I'm using the quickest approach to getting all requested changes made + everything rebased. Feel free to refactor it or create a ticket to clean this up but it's not something I'm going to be able to prioritize

driusan commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -269,7 +269,67 @@ class RequestAccount extends \NDB_Form
+        //     - Associated with the Site
+        //     - Associated with the Project
+        //     - Have the User Management permission
+        $query  = <<<SQL
+        SELECT DISTINCT u.Email,u.userID 
+        FROM users u 
+        JOIN user_psc_rel upr 
+        ON u.ID = upr.UserID 
+        JOIN user_project_rel uprojr 
+        ON u.ID = uprojr.UserID 
+        WHERE upr.CenterID=:center
+        AND uprojr.ProjectID=:project
+SQL;
```
```suggestion
        $query  = "SELECT DISTINCT u.Email,u.userID
             FROM users u
                JOIN user_psc_rel upr
                    (ON u.ID = upr.UserID)
               JOIN user_project_rel uprojr
                   (ON u.ID = uprojr.UserID)
            WHERE upr.CenterID=:center
                AND uprojr.ProjectID=:project"
```

Formatting SQL isn't a good use of HEREDOCs and also the formatting is hard to read.

ridz1208 commented:
@driusan 

@johnsaigle 's explanations make sense. your call

driusan commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -271,5 +271,46 @@ class RequestAccount extends \NDB_Form
+            $config    = \NDB_Factory::singleton()->config();
+            $studyName = $config->getSetting('title');
+
+            \Email::send(
```
This is still missing a sandbox check

cmadjar replied:
unless I am misreading this, I believe the sandbox check is done line 276 (returns from the function and this code is not executed).

cmadjar replied:
nevermind, I guess this was resolved before John left. Will resolve this comment.

johnsaigle commented:
@driusan I added the sandbox check.

I also reverted your MySQL changes -- the caused a syntax error on the back-end. I incorporated your indentation though to help with readability.

johnsaigle commented:
@driusan could you merge if satisfied?

driusan commented on modules/login/php/requestaccount.class.inc:
```diff
@@ -271,5 +271,53 @@ class RequestAccount extends \NDB_Form
+        //     - Associated with the Site
+        //     - Associated with the Project
+        //     - Have the User Management permission
+        $query  = <<<SQL
```
This still needs to be changed from a HEREDOC to normal PHP quotes by whoever picks this up, then it'll be ready to be merged.

cmadjar commented:
@driusan ready for review and merge :)
